light-system-programs/programs/merkle_tree_program/src/lib.rs

```diff
@@ -90,35 +92,31 @@ pub mod merkle_tree_program {
 
         new_merkle_tree.lock_duration = lock_duration;
 
-        Ok(())
-    }
+        // Event Merkle Tree
 
-    pub fn initialize_new_event_merkle_tree(
-        ctx: Context<InitializeNewEventMerkleTree>,
-    ) -> Result<()> {
-        if !ctx
-            .accounts
-            .merkle_tree_authority_pda
-            .enable_permissionless_merkle_tree_registration
-            && ctx.accounts.authority.key() != ctx.accounts.merkle_tree_authority_pda.pubkey
-        {
-            return err!(ErrorCode::InvalidAuthority);
+        let old_merkle_tree_loader: AccountLoader<EventMerkleTree> =
+            AccountLoader::try_from(&ctx.remaining_accounts[1])?;
+        let old_merkle_tree_key = old_merkle_tree_loader.key();
+        let mut old_merkle_tree = old_merkle_tree_loader.load_mut()?;
+
+        let index = old_merkle_tree.merkle_tree_nr;
+        let (pubkey, _) = Pubkey::find_program_address(
+            &[EVENT_MERKLE_TREE_SEED, index.to_le_bytes().as_ref()],
+            ctx.program_id,
+        );
+        if old_merkle_tree_key != pubkey {
+            return err!(ErrorCode::InvalidOldMerkleTree);
```

Comment:
we have some duplicate code here:

https://github.com/Lightprotocol/light-protocol/pull/227/files#diff-103721ed49551c60e6e7f547823b334a3d789e8f379695eb02b318c908a67f12L62-L82

https://github.com/Lightprotocol/light-protocol/pull/227/files#diff-103721ed49551c60e6e7f547823b334a3d789e8f379695eb02b318c908a67f12R97-R114

we could just loop over the remaining accounts

Reply:
What do you mean by "loop over the remaining accounts"?

First of all, we expect concrete types at concrete indexes and there are just 2 of them, so I don't see how a loop could be useful.

I agree that the code looks duplicate, but TransactionMerkleTree and EventMerkleTree are different structs. I don't see modifying their fields could be deduplicated without making a macro (overkill IMO) or writing a trait for them (even bigger overkill IMO).

I could try to make it less duplicate by making some util function for deserializing and checking seeds, which takes an account type as a generic, so at least we get rid of repeating

```rust
        let old_merkle_tree_loader: AccountLoader<EventMerkleTree> =
            AccountLoader::try_from(&ctx.remaining_accounts[1])?;
        let old_merkle_tree_key = old_merkle_tree_loader.key();
        let mut old_merkle_tree = old_merkle_tree_loader.load_mut()?;

        let index = old_merkle_tree.merkle_tree_nr;
        let (pubkey, _) = Pubkey::find_program_address(
            &[EVENT_MERKLE_TREE_SEED, index.to_le_bytes().as_ref()],
            ctx.program_id,
        );
        if old_merkle_tree_key != pubkey {
            return err!(ErrorCode::InvalidOldMerkleTree);
        }
```

Let me do that in the next push and then we can discuss if that's enough or whether we should proceed with macro/trait.

Reply:
Actually, I tried and it looks like it's impossible to do any deduplication without a trait.

Reply:
Done. At the end I made a small macro and a trait, lol. I couldn't figure out any better solution. Let me know if you agree with it https://github.com/Lightprotocol/light-protocol/pull/227/commits/90874ecb16af509f97fdd7411c2a875a7e2d8b18

Reply:
my idea was sth like this:
```
for (i, account) in ctx.remaining_accounts.iter().enumerate() {
            if i == 0 {
                // Transaction Merkle Tree
                checkMerkleTree<TransactionMerkleTree,Context<InitializeNewMerkleTrees>>(&ctx.remaining_accounts[0], TRANSACTION_MERKLE_TREE_SEED, ctx)?;

            } else if i == 1 {
                checkMerkleTree<TransactionMerkleTree,Context<InitializeNewMerkleTrees>>(&ctx.remaining_accounts[1], EVENT_MERKLE_TREE_SEED, ctx)?;
            } else {
                return err!(ErrorCode::ExpectedOldMerkleTrees);
            }
        }
        
```
```
pub fn checkMerkleTree<T: Account<'_>, C: Context>(
    accountInfo: AccountInfo<'_>,
    seed: str,
    ctx: &C,
) -> Result<()> {
    deserialize_and_update_old_merkle_tree::<TransactionMerkleTree>(
        &accountInfo,
        seed,
        ctx.program_id,
    )?;
    let old_merkle_tree_loader: AccountLoader<T> = AccountLoader::try_from(&accountInfo)?;
    let old_merkle_tree_key = old_merkle_tree_loader.key();
    let mut old_merkle_tree = old_merkle_tree_loader.load_mut()?;

    let index = old_merkle_tree.merkle_tree_nr;
    let (pubkey, _) = Pubkey::find_program_address(
        &[TRANSACTION_MERKLE_TREE_SEED, index.to_le_bytes().as_ref()],
        ctx.program_id,
    );
    if old_merkle_tree_key != pubkey {
        return err!(ErrorCode::InvalidOldMerkleTree);
    }

    if old_merkle_tree.newest != 1 {
        return err!(ErrorCode::NotNewestOldMerkleTree);
    }
    old_merkle_tree.newest = 0;

    Ok(())
}
```

Reply:
Just seeing that the loop is not necessary

Reply:
nice! Yeah this is what I had in mind, just didn't think we need a trait.